src/humanPose/Animation.js

```diff
@@ -0,0 +1,142 @@
+const AnimationState = {
+    video: 'video',
+    noVideo: 'noVideo',
+};
+
+/**
+ * Animation
+ * Organizes all the information about an HPA's animation in one location
+ * Coordinates propagating animation information to the motionStudy and its
+ * video player if present
+ *
+ * Notably hands off control of video playback to that video player during any
+ * region of the recording that it contains
+ *
+ * Time is in ms unless otherwise specified
+ */
+export class Animation {
+    constructor(humanPoseAnalyzer, motionStudy, startTime, endTime) {
+        this.humanPoseAnalyzer = humanPoseAnalyzer;
+        this.motionStudy = motionStudy;
+
+        this.startTime = startTime;
+        this.endTime = endTime;
+
+        this.cursorTime = this.startTime;
+        this.lastUpdate = -1;
+        this.playing = true;
+
+        this.animationState = AnimationState.noVideo;
+
+        if (this.videoPlayer) {
+            this.attachVideoPlayerEvents();
+        }
+    }
+
+    get videoPlayer() {
+        return this.motionStudy.videoPlayer;
+    }
+
+    get videoStartTime() {
+        return this.motionStudy.videoStartTime;
+    }
+
+    attachVideoPlayerEvents() {
+        const video = this.videoPlayer.colorVideo;
+        video.addEventListener('play', () => {
+            this.playing = true;
+        });
+        video.addEventListener('pause', () => {
+            if (this.animationState !== AnimationState.video) {
+                return;
+            }
+
+            if (this.humanPoseAnalyzer.active) {
+                this.playing = false;
+            } else if (this.playing) {
+                this.videoPlayer.play();
+            }
+        });
+    }
+
+    update(time) {
+        if (this.lastUpdate < 0) {
+            this.lastUpdate = time;
+        }
+        let dt = time - this.lastUpdate;
+        this.lastUpdate = time;
+        const duration = this.endTime - this.startTime;
+        if (this.playing) {
+            this.cursorTime += dt;
+            if (this.cursorTime < this.startTime) {
+                this.cursorTime = this.startTime;
+            }
+        }
+        let offset = this.cursorTime - this.startTime;
+        if (offset > duration) {
+            offset = 0;
+            this.cursorTime = this.startTime;
+            // Reset animation state
+            if (this.animationState === AnimationState.video) {
+                this.stopVideoPlayback();
+            }
+        }
+
+        // Adjusted cursorTime that might be tracking a
+        // wobbly-relative-to-real-time video playback
+        let cursorTimeAdj = this.cursorTime;
+        if (this.isVideoPlayerInControl(cursorTimeAdj) && this.playing) {
+            if (this.animationState !== AnimationState.video) {
+                this.startVideoPlayback(cursorTimeAdj);
+            } else {
+                cursorTimeAdj = this.cursorTimeFromVideo();
```

Comment:
So if I understand correctly, video player does equivalent of 'this.cursorTime += dt;' and the calculation here is overriden.

Reply:
This Animation does `this.cursorTime += dt`. The VideoPlayer uses its color video element's currentTime property which exactly tracks where the video is.  This code `this.cursorTime` with the more-exact video-derived cursorTimeFromVideo() for pose display. It doesn't write to `this.cursorTime` to guarantee that the animation continues eventually. For example, if the video starts buffering and pauses, then `this.cursorTimeFromVideo()` will stop moving forwards as well. Not writing guarantees that the loop will eventually start from the beginning again